site/content/features/virtual-threads.md

```diff
@@ -0,0 +1,427 @@
+import java.net.*;
+import java.net.http.*;
+
+public class RateLimitDemo {
+   public static void main(String[] args) throws InterruptedException, ExecutionException {
+      ExecutorService service = Executors.newVirtualThreadPerTaskExecutor();
+      List<Future<String>> futures = new ArrayList<>();
+      final int TASKS = 250;
+      for (int i = 1; i <= TASKS; i++)
+         futures.add(service.submit(() -> get("https://horstmann.com/random/word")));
+      for (Future<String> f : futures)
+         System.out.print(f.get() + " ");
+      System.out.println();
+      service.close();
+   }
+
+   private static HttpClient client = HttpClient.newHttpClient();
+
+   private static final Semaphore SEMAPHORE = new Semaphore(20);
+   
+   public static String get(String url) {
+      try {
+         var request = HttpRequest.newBuilder().uri(new URI(url)).GET().build();
+         SEMAPHORE.acquire();
+         try {
+            Thread.sleep(100);
+            return client.send(request, HttpResponse.BodyHandlers.ofString()).body();
+         } finally {
+            SEMAPHORE.release();
+         }
+      } catch (Exception ex) {
+         ex.printStackTrace();
+         var rex = new RuntimeException();
+         rex.initCause(ex);
+         throw rex;
+      }
+   }   
+}
+
+{{< /sandboxsource >}}
+{{< /sandbox >}}
+
+## Pinning
+
+The virtual thread scheduler mounts virtual threads onto carrier threads. By default, there are as many carrier threads as there are CPU cores. You can tune that count with the `jdk.virtualThreadScheduler.parallelism` VM option.
+
+When a virtual thread executes a blocking operation, it is supposed to be unmounted from its its carrier thread, which can then execute a different virtual thread. There are two situations in which this does *not* happen:
+
+1. When executing a `synchronized` method or block
+2. When calling a native method or foreign function
+
+A thread that does either of  these is called *pinned*. Being pinned is not bad in itself. But when a pinned thread blocks, there is a problem. The thread cannot be unmounted, and the carrier thread is blocked. That leaves fewer carrier threads for running virtual threads.
+
+In some situations, the virtual thread scheduler will compensate by starting another carrier thread. For example, in JDK 21, this happens for many file I/O operations, and when calling `Object.wait`. But this is an implementation detail that may change. You can control the maximum number of carrier threads with the `jdk.virtualThreadScheduler.maxPoolSize` VM option.
+
+Pinning is harmless if `synchronized` is used to avoid a race condition in an in-memory operation. However, if there are blocking calls, it would be best to replace `synchronized` with a `ReentrantLock`. This is of course only an option if you have control over the source code. 
+
+You always have the option whether or not to use virtual threads. In particular, heavy file I/O work may be better suited for platform threads.
```

Comment:
Why is file I/O different to network I/O?

Reply:
Because network I/O will reliably unmount the virtual thread. The interplay of unmounting, file I/O, and pinning is messy. I am going to push a rewritten section on pinning to hopefully disentangle it better, and I explicitly say that network I/O will unmount (except when pinning...)

Reply:
Thanks for the clarification! I like the reworded version.